File: app/services/excel_parser.py
"""Excel file parser for financial statement data."""
from typing import List, Dict, Tuple, Any, Optional
from decimal import Decimal, InvalidOperation
import openpyxl
from openpyxl.utils import get_column_letter, column_index_from_string


def parse_excel_file(file_path: str) -> List[Dict[str, Any]]:
    """
    Parse Excel file and extract financial data rows.
    
    Args:
        file_path: Path to .xlsx file
        
    Returns:
        List of dictionaries, each representing a row with keys:
        - item_name (str)
        - category (str)
        - amount (Decimal or None if invalid)
        - notes (str or None)
        - row_number (int): Excel row number (1-indexed)
        
    Raises:
        FileNotFoundError: If file doesn't exist
        ValueError: If file is not a valid Excel file
    """
    try:
        workbook = openpyxl.load_workbook(file_path, data_only=True)
        sheet = workbook.active  # Use first sheet
        
        rows = []
        header_row = None
        header_map = {}  # Maps column letter to field name
        
        # Find header row (first non-empty row)
        for row_idx, row in enumerate(sheet.iter_rows(min_row=1, max_row=100, values_only=False), start=1):
            values = [cell.value for cell in row if cell.value is not None]
            if values:
                # Check if this looks like a header row
                header_candidates = [str(v).lower().strip() for v in values]
                if any(keyword in ' '.join(header_candidates) for keyword in ['item', 'category', 'amount', 'name']):
                    header_row = row_idx
                    # Map columns to field names
                    for col_idx, cell in enumerate(row, start=1):
                        if cell.value:
                            col_letter = get_column_letter(col_idx)
                            value_lower = str(cell.value).lower().strip()
                            # Map common column name variations
                            if 'item' in value_lower and 'name' in value_lower:
                                header_map[col_letter] = 'item_name'
                            elif 'name' in value_lower and 'item' not in value_lower:
                                header_map[col_letter] = 'item_name'
                            elif 'category' in value_lower:
                                header_map[col_letter] = 'category'
                            elif 'amount' in value_lower or 'value' in value_lower:
                                header_map[col_letter] = 'amount'
                            elif 'note' in value_lower or 'comment' in value_lower or 'description' in value_lower:
                                header_map[col_letter] = 'notes'
                    break
        
        if not header_row:
            # Fallback: assume first row is header with standard column order
            # item_name, category, amount, notes
            header_map = {'A': 'item_name', 'B': 'category', 'C': 'amount', 'D': 'notes'}
            header_row = 1
        
        # Parse data rows
        for row_idx, row in enumerate(sheet.iter_rows(min_row=header_row + 1, values_only=False), start=header_row + 1):
            row_data = {}
            has_data = False
            
            for col_letter, field_name in header_map.items():
                col_idx = column_index_from_string(col_letter)
                if col_idx <= len(row):
                    cell = row[col_idx - 1]
                    value = cell.value
                    
                    if field_name == 'amount':
                        # Try to parse as number
                        if value is None:
                            row_data[field_name] = None
                        elif isinstance(value, (int, float)):
                            try:
                                row_data[field_name] = Decimal(str(value))
                            except (InvalidOperation, ValueError):
                                row_data[field_name] = None
                        else:
                            # Try to parse string as number
                            try:
                                # Remove commas and whitespace
                                clean_value = str(value).replace(',', '').replace(' ', '').strip()
                                row_data[field_name] = Decimal(clean_value)
                            except (InvalidOperation, ValueError):
                                row_data[field_name] = None
                    else:
                        # String fields
                        if value is not None:
                            row_data[field_name] = str(value).strip()
                            if row_data[field_name]:
                                has_data = True
                        else:
                            row_data[field_name] = None
                else:
                    row_data[field_name] = None
            
            # Only include row if it has at least item_name or category
            if has_data or row_data.get('item_name') or row_data.get('category'):
                row_data['row_number'] = row_idx
                rows.append(row_data)
        
        workbook.close()
        return rows
        
    except FileNotFoundError:
        raise FileNotFoundError(f"Excel file not found: {file_path}")
    except Exception as e:
        raise ValueError(f"Error parsing Excel file: {str(e)}")


def validate_excel_structure(rows: List[Dict[str, Any]]) -> Tuple[bool, List[str]]:
    """
    Validate that parsed rows have required structure.
    
    Args:
        rows: List of parsed row dictionaries
        
    Returns:
        Tuple of (is_valid: bool, errors: List[str])
    """
    errors = []
    
    if not rows:
        errors.append("Excel file contains no data rows")
        return False, errors
    
    # Check for required fields in at least some rows
    has_item_name = any(row.get('item_name') for row in rows)
    has_category = any(row.get('category') for row in rows)
    has_amount = any(row.get('amount') is not None for row in rows)
    
    if not has_item_name:
        errors.append("Missing required column: item_name")
    if not has_category:
        errors.append("Missing required column: category")
    if not has_amount:
        errors.append("Missing required column: amount")
    
    is_valid = has_item_name and has_category and has_amount
    return is_valid, errors

File: frontend/src/utils/excelParser.js
/**
 * Excel Parser Utility
 * 
 * Client-side Excel file parsing using SheetJS (xlsx)
 */

import * as XLSX from 'xlsx';

/**
 * Parse Excel file and extract financial data rows
 * @param {File} file - Excel file (.xlsx)
 * @returns {Promise<Array>} Array of row objects with item_name, category, amount, notes, row_number
 */
export async function parseExcelFile(file) {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    
    reader.onload = (e) => {
      try {
        const data = new Uint8Array(e.target.result);
        const workbook = XLSX.read(data, { type: 'array' });
        const sheetName = workbook.SheetNames[0];
        const sheet = workbook.Sheets[sheetName];
        
        // Convert to JSON with header row
        const jsonData = XLSX.utils.sheet_to_json(sheet, { 
          header: 1, // Use first row as header
          defval: null, // Default value for empty cells
          raw: false // Convert numbers to strings for processing
        });
        
        if (jsonData.length === 0) {
          reject(new Error('Excel file is empty'));
          return;
        }
        
        // Find header row (first non-empty row)
        let headerRowIndex = 0;
        let headerMap = {};
        
        // Try to find header row
        for (let i = 0; i < Math.min(jsonData.length, 10); i++) {
          const row = jsonData[i];
          if (row && row.length > 0) {
            const rowValues = row.filter(cell => cell !== null && cell !== undefined && String(cell).trim() !== '');
            if (rowValues.length >= 2) {
              // Check if this looks like a header row
              const rowText = rowValues.map(cell => String(cell).toLowerCase()).join(' ');
              if (rowText.includes('item') || rowText.includes('category') || rowText.includes('amount') || rowText.includes('name')) {
                headerRowIndex = i;
                // Map columns to field names
                row.forEach((cell, colIndex) => {
                  if (cell) {
                    const cellLower = String(cell).toLowerCase().trim();
                    if ((cellLower.includes('item') && cellLower.includes('name')) || (cellLower.includes('name') && !cellLower.includes('category'))) {
                      headerMap[colIndex] = 'item_name';
                    } else if (cellLower.includes('category')) {
                      headerMap[colIndex] = 'category';
                    } else if (cellLower.includes('amount') || cellLower.includes('value')) {
                      headerMap[colIndex] = 'amount';
                    } else if (cellLower.includes('note') || cellLower.includes('comment') || cellLower.includes('description')) {
                      headerMap[colIndex] = 'notes';
                    }
                  }
                });
                break;
              }
            }
          }
        }
        
        // If no header map found, use default order: item_name, category, amount, notes
        if (Object.keys(headerMap).length === 0) {
          headerMap = { 0: 'item_name', 1: 'category', 2: 'amount', 3: 'notes' };
        }
        
        // Parse data rows
        const rows = [];
        for (let i = headerRowIndex + 1; i < jsonData.length; i++) {
          const row = jsonData[i];
          if (!row || row.length === 0) continue;
          
          const rowData = {};
          let hasData = false;
          
          // Extract data based on header map
          Object.entries(headerMap).forEach(([colIndex, fieldName]) => {
            const colIdx = parseInt(colIndex);
            if (colIdx < row.length && row[colIdx] !== null && row[colIdx] !== undefined) {
              const value = row[colIdx];
              
              if (fieldName === 'amount') {
                // Try to parse as number
                if (typeof value === 'number') {
                  rowData[fieldName] = value;
                  hasData = true;
                } else if (typeof value === 'string') {
                  // Remove commas and whitespace, then parse
                  const cleanValue = value.replace(/,/g, '').replace(/\s/g, '').trim();
                  const numValue = parseFloat(cleanValue);
                  if (!isNaN(numValue)) {
                    rowData[fieldName] = numValue;
                    hasData = true;
                  } else {
                    rowData[fieldName] = null;
                  }
                } else {
                  rowData[fieldName] = null;
                }
              } else {
                // String fields
                const strValue = String(value).trim();
                if (strValue) {
                  rowData[fieldName] = strValue;
                  hasData = true;
                } else {
                  rowData[fieldName] = null;
                }
              }
            } else {
              rowData[fieldName] = null;
            }
          });
          
          // Only include row if it has at least item_name or category
          if (hasData || rowData.item_name || rowData.category) {
            rowData.row_number = i + 1; // Excel row number (1-indexed)
            rows.push(rowData);
          }
        }
        
        resolve(rows);
      } catch (error) {
        reject(new Error(`Error parsing Excel file: ${error.message}`));
      }
    };
    
    reader.onerror = () => {
      reject(new Error('Failed to read file'));
    };
    
    reader.readAsArrayBuffer(file);
  });
}

/**
 * Validate that parsed rows have required structure
 * @param {Array} rows - Array of parsed row objects
 * @returns {{valid: boolean, errors: Array<string>}} Validation result
 */
export function validateExcelStructure(rows) {
  const errors = [];
  
  if (!rows || rows.length === 0) {
    errors.push('Excel file contains no data rows');
    return { valid: false, errors };
  }
  
  // Check for required fields in at least some rows
  const hasItemName = rows.some(row => row.item_name);
  const hasCategory = rows.some(row => row.category);
  const hasAmount = rows.some(row => row.amount !== null && row.amount !== undefined);
  
  if (!hasItemName) {
    errors.push('Missing required column: item_name');
  }
  if (!hasCategory) {
    errors.push('Missing required column: category');
  }
  if (!hasAmount) {
    errors.push('Missing required column: amount');
  }
  
  const valid = hasItemName && hasCategory && hasAmount;
  return { valid, errors };
}
